### ISSUE TITLE
Handling partial DAGs in Web3.Storage

### ISSUE THREAD

**Issue by dchoi27** [web3-storage/web3.storage#394]
Currently, Web3.Storage does not handle partial DAGs well. The pin status of a partial DAG will never show as complete, and the Filecoin deal aggregator does not support partial DAGs (+ there are problems storing / retrieving partial DAGs in Filecoin). However, there will be cases where partial DAGs are uploaded to Web3.Storage both unintentionally (e.g., user connection interrupted during upload) or intentionally (e.g., database transactions, diffs in general). We should figure out how we handle these cases.

cc @olizilla @mikeal if anything to add!

**Comment by lidel** [comment]
Is the underlying problem here that blocks that are not in a CAR may not be in web3.storage's local repo and  pinning is stuck at finding them, or something else?

My thinking: if I add a complete DAG, and then keep submitting CARs with delta that reuses data already uploaded before, this would not require any external fetch, because web3.storage's repo should (in theory) have all necessary blocks.

**Comment by dchoi27** [comment]
Yeah I think that's basically it - right now the infra to get things onto Filecoin gets stuck if it's a partial since it's trying to fetch the pin, and Cluster never can report pin status as complete. I think the problem generally arises because uploads happen via chunked CAR files, but if this terminates mid-way through, then on the server side we don't actually know it's a partial, but users often don't realize that their upload didn't finish. But just generally handling partials better will unlock new use cases as well. @mikeal is this right?
